### PR TITLE
Higher rank Arbitrary and Coarbitrary classes

### PR DESCRIPTION
**Pull request by garyb**
I'm looking at QC-laws again, it seems these classes would be useful generically, not just in there.

Maybe should port over `eq1` and `eq2` too?

The `arbArbitrary1` and `arb1Arbitrary2` give us `Arbitrary` and `Arbitrary1` instance for every `Arbitrary2` "for free" and won't result in overlaps as long as no attempt is made to implement the lower-rank versions for a type that has the higher-rank instance.

### CODE REVIEW

**Comment by paf31** [comment]
How do you see these being used elsewhere, where we couldn't use the existing versions?

I think I'd prefer to reimplement the overlap check for instances at the definition site before adding flexible instances like these. That way, with your new check for orphans, they'd be safe, right?

**Comment by garyb** [comment]
It already fails at the definition site if you define an overlapping instance :) I was pleasantly surprised by that.

Law checks generally would be the use case for them, not just my core laws package. I went back to looking at your PR and really liked it, until I realised I would have to reimplement a ton of things as higher rank `Arbitrary1`s, which seemed silly since we could just provide them in the first place and do the instance trick here to get the other varieties at the same time.

**Comment by garyb** [comment]
```
Error at D:\Home\GitHub\purescript-quickcheck\src\Test\QuickCheck\Arbitrary.purs line 117, column 1 - line 122, column 1:
  Overlapping instances found for Test.QuickCheck.Arbitrary.Arbitrary (Prim.Array a248):
  Test.QuickCheck.Arbitrary.arbArbitrary1 via
    Test.QuickCheck.Arbitrary.arb1Array
    Dictionary in scope
  Test.QuickCheck.Arbitrary.arbArray via
    Dictionary in scope
```

File: src/Test/QuickCheck/Arbitrary.purs [hunk 1]
```diff
@@ -20,6 +20,18 @@ import Test.QuickCheck.Gen
 class Arbitrary t where
   arbitrary :: Gen t
 
+class Arbitrary1 f where
```
Comment (paf31) [inline src/Test/QuickCheck/Arbitrary.purs]: This is `Lifting Arbitrary f` in `constraints

**Pull request closed without merging**